### PR TITLE
ssh-agent module: run in foreground as upstream recommends

### PR DESCRIPTION
**Pull request by peterhoeg**
I'm using it here and it works perfectly fine.

Systemd's official recommendation is that everything should run in the foreground.
###### Things done
- [X] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [X] NixOS
  - [ ] OS X
  - [ ] Linux
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by edolstra** [comment]
Things should only run in the foreground if they provide startup notification to systemd. Otherwise the unit will reach "started" state despite not being ready. (Of course the unit may not be ready if it forks either, but that's generally the convention.) Does ssh-agent do startup notification?

**Comment by peterhoeg** [comment]
Thank you for the quick feedback @edolstra. ```ssh-agent``` does not use ```sd_notify``` to signal readiness (which would change the type from "simple" to "notify").

This issue is similar to the thread on #14881 - the same arguments on that PR would apply here.

If we keep it ```forking``` shouldn't we review all the other services that are set to ```simple``` or blank to verify that the ```forking``` approach actually finishes initializing before forking and then changing them to ```forking``` instead?

**Comment by joachifm** [comment]
@peterhoeg such a review makes sense. If there are `simple` services that are not actually ready upon `exec()`, that's a bug waiting to happen. Using `simple` wherever possible makes sense, but we must take care not to break unit ordering.

Waiting for the ssh-agent socket to appear means the unit will at least be closer to a true "ready" state, but I'm not familiar enough with ssh-agent to say whether that's acceptable. It's more code to accomplish what the previous definition did out of the box, though.

File: nixos/modules/programs/ssh.nix [hunk 1]
```diff
@@ -204,11 +204,12 @@ in
         serviceConfig =
           { ExecStartPre = "${pkgs.coreutils}/bin/rm -f %t/ssh-agent";
             ExecStart =
-                "${cfg.package}/bin/ssh-agent " +
+                "${cfg.package}/bin/ssh-agent -D " +
                 optionalString (cfg.agentTimeout != null) ("-t ${cfg.agentTimeout} ") +
                 "-a %t/ssh-agent";
+            ExecStartPost = "/bin/sh -c 'set -eu ; while [ ! -S $SSH_AUCK_SOCK ]; do sleep 1; done'";
```
Comment (joachifm) [inline nixos/modules/programs/ssh.nix]: I wonder, isn't sleeping for a whole second between tests a bit excessive? `sleep` accepts fractional arguments.

**Comment by peterhoeg** [comment]
I was thinking about this earlier, and in general (regardless of ```simple``` or ```forking```), wouldn't it make sense to ship an extremely small helper binary that can either probe a port, socket or check for the presence of a file and simply ```exit(0);``` when it appears and just sleep in-between in increments. 

This could be executed as a Post action for pretty much all the services as without a code review for every service, there is really no way to know if a particular service is ready.

The other way to approach this is to simply say - I don't care. Services can come and go, so if you are depending on a service, you also need to gracefully be able to deal with it not being there and simply try again a bit later. Systemd should be restarting any failed services, but there will be a window where that service is not ready. I must admit, I'm leaning towards this...

**Comment by joachifm** [comment]
Inferring a daemon's internal state from the existence of files or sockets seems brittle, in general. It seems impossible to get around knowing a fair bit about the daemon the service unit is executing, and this proposal seems to *increase* the need to know what the daemon is doing under the hood. Being ready upon `fork()` is at least idiomatic, not sure the same can be said about "this file exists". 

For existing `simple` services, adding `ExecStartPost` to ensure readiness using your proposed helper to avoid boilerplate sounds like a good plan, but it's not quite clear to me what is gained by doing so for `forking` services, except in cases where the daemon forks "just because" and can be made not to.

**Comment by peterhoeg** [comment]
Just because something is forking, doesn't mean it's ready. And a standardized way to at least let us WaitForSomething(tm) would make things easier to maintain.

Do you have an example of units where we absolutely can only continue if the service is actually ready?

**Comment by joachifm** [comment]
> Just because something is forking, doesn't mean it's ready.

Sure, I still hold that you need to know quite a lot about what a daemon is doing to sanely infer anything from the existence of files, whereas `fork()` is a long-standing convention.

> Do you have an example of units where we absolutely can only continue if the service is actually ready?

No. The issues I typically encounter have to do with ordering, not readiness. Most of the services I use don't appear to expect anything more than "this process is running and will eventually respond".

**Comment by peterhoeg** [comment]
We basically have the following options when it comes to services and readiness:

1. Socket activation
2. Readiness signalling via ```sd_notify```
3. Readiness signalling by taking a dbus ```BusName```
4. Double forking, where the ```forking``` only happens when init is done
5. ```simple``` services where we don't care about state

Of those 5 options only the first is guaranteed to work as the kernel will buffer connections to the socket.

Options 2 and 3 will indicate the service is up, but will not make a difference if the service goes down.

Option 4 *may* only continue if the service is up, but there are no guarantees.

Option 5 - whatever basically.


But the only time this actually makes a difference in the way things are currently implemented is during startup. If a dependent service goes down, anything depending on that STILL needs to handle that situation.

So us adding ```ExecStartPost``` only deals with with startup scenario anyway, which is why I find the whole ordeal a little pointless.

**Comment by joachifm** [comment]
@peterhoeg I agree that readiness without an explicit protocol is hopeless in general. I disagree that the potential for run-time failures invalidates readiness, though. My understanding is that readiness mainly factors into the creation of optimal activation plans. If "ready" doesn't mean "actually ready", then invalid plans may result --- embracing the systemd way means caring about readiness. Anyway, I don't think there's much point in taking this further here (I must confess that I find the topic as such more interesting than this PR itself, I don't mean to be difficult).

**Comment by edolstra** [comment]
Closing this because it's not an improvement. `Type=simple` is worse than `Type=forking` if the latter forks after reaching readiness, which is the case for `ssh-agent` (I just checked the source). In fact `Type=simple` is almost never a good idea because of the race condition it introduces. The critical postcondition of a systemd service is that when it reached "started" stage, it's ready to accept connections. Otherwise units or scripts that depend on the service will fail randomly, e.g.
```
systemctl --user start ssh-agent
ssh-add ...
```
With `Type=simple`, the latter call will fail randomly, so you'd have to introduce a loop to wait for readiness. And of course that loop will hang forever if the service dies in the meantime.

What *would* be an improvement is to add a one-line patch to ssh-agent to call `sd_notify()` at the appropriate time, or (more complicated) support socket activation.

**Pull request closed without merging**